//define syntax
syntax = "proto3";

package sss;

//define buyer services
service AllServices{

    //search item
    rpc SearchItem(SearchItemRequest) returns (SearchItemResponse);

    //buy item
    rpc BuyItem(BuyItemRequest) returns (BuyItemResponse);

    //add to wish list
    rpc AddToWishList(AddToWishListRequest) returns (AddToWishListResponse);

    //rate item
    rpc RateItem(RateItemRequest) returns (RateItemResponse);

    //register seller
    rpc RegisterSeller(RegisterSellerRequest) returns (RegisterSellerResponse);

    //sell item
    rpc SellItem(SellItemRequest) returns (SellItemResponse);

    //update item
    rpc UpdateItem(UpdateItemRequest) returns (UpdateItemResponse);

    //delete item
    rpc DeleteItem(DeleteItemRequest) returns (DeleteItemResponse);

    //display items
    rpc DisplaySellerItems(DisplaySellerItemsRequest) returns (DisplaySellerItemsResponse);

    //notify client
    rpc NotifyClient(Notification) returns (NotificationResponse);
}

//define item structure
message Item{
    int32 item_id = 1;
    string name = 2;
    float item_price = 3;
    int32 quantity = 4;
    string category = 5;
    string description = 6;
    string seller_address = 7;
    float rating = 8;
}

//define seller structure
message Seller {
    string seller_id = 1;
    string name = 2;
}

//define notification
message Notification {
  string message = 1;
}

//define request and response messages
message NotificationResponse {
  string message = 1;
}

//define request and response messages
message RegisterSellerRequest{
    string message = 1;
}

message RegisterSellerResponse{
    string message = 1;
}

message SellItemRequest{
    string message = 1;
    Item item = 2;
    string seller_id = 3;
}

message SellItemResponse{
    string message = 1;
}

message UpdateItemRequest{
    string seller_id = 1;
    int32 item_id = 2;
    float new_item_price = 3;
    int32 new_quantity = 4;
    string new_seller_address = 5;
}

message UpdateItemResponse{
    string message = 1;
}

message DeleteItemRequest{
    string seller_id = 1;
    string seller_address = 2;
    int32 item_id = 3;
}

message DeleteItemResponse{
    string message = 1;
}

message DisplaySellerItemsRequest{
    string seller_id = 1;
}

message DisplaySellerItemsResponse{
    repeated Item items = 1;
}

//define request and response messages
message SearchItemRequest{
    string item_name = 1;
    string category_name = 2;
}

message SearchItemResponse{
    repeated Item items = 1;
}

message BuyItemRequest{
    int32 item_id = 1;
    int32 item_quantity = 2;
}

message BuyItemResponse{
    string status = 1;
}

message AddToWishListRequest{
    int32 item_id = 1;
}

message AddToWishListResponse{
    string status = 1;
}

message RateItemRequest{
    int32 item_id = 1;
    int32 rating = 2;
}

message RateItemResponse{
    string status = 1;
}